pandas/io/json/normalize.py

```diff
@@ -224,6 +224,44 @@ def _pull_field(js, spec):
         sep = str(sep)
     meta_keys = [sep.join(val) for val in meta]
 
+    def _extract(obj, field, seen_meta, level):
+        """
+        Extract field from obj and update `records`, `lengths`, `meta_vals`.
+
+        Parameters
+        ----------
+        obj : dict
+            Unserialized dict
+        field : str
+            The field to extract from obj
+        seen_meta : dict
+            The dict of meta values that have been visited
+        level: int
+            The current level
+        """
+        recs = _pull_field(obj, field)
+
+        # For repeating the metadata later
+        lengths.append(len(recs))
+
+        for val, key in zip(meta, meta_keys):
```

Comment:
can you make this  immutable function, this is very hard to reason about now

Reply:
`_extract` is a new helper function. The original function `_recursive_extract` (and now the caller of `_extract`) is not immutable.
It's strange that we make `_extract` immutable but `_recursive_extract` is not. 
In this PR I want to make minimal changes. Refactoring it is certainly not in scope of this PR.

Reply:
@vuminhle if you are going to move it, then i'd like to fix it. its not a problem to also fix ``_recursive_extract``.

Reply:
Agreed with @jreback here; would rather see a change that addresses some of the mutability concerns with `recursive_extract` rather than creating a new function

Reply:
I refactor `_recursive_extract` to make it return a tuple of `records` and `meta_vals` (`lengths` was not necessary because it could be computed from `records`).